native/exe/SageLauncher/launcher.cpp

```diff
@@ -993,63 +1014,70 @@ int launchJVMSage(LPSTR lpszCmdLine, HWND hWnd, BOOL bClient, BOOL bService)
+	LPTSTR libraryPath = new TCHAR[sizelibraryPath];
 	
 	stdOutHandle = GetStdHandle(STD_OUTPUT_HANDLE);
     JNIEnv *env;       /* pointer to native method interface */
 	JavaVMInitArgs vm_args;
 	JavaVMOption options[32];
 	vm_args.nOptions = 0;
 #ifdef SAGE_TV
-	strcpy(prefFile, bClient ? "sagetvclient " : "sagetv ");
-	strcat(prefFile, appPath);
-	strcat(prefFile, bClient ? "SageClient.properties" : "Sage.properties");
-	strcpy(jarPath, "-Djava.class.path=");
-	strcat(jarPath, appPath);
-	strcat(jarPath, "Sage.jar;");
-  strcat(jarPath, appPath);
-  strcat(jarPath, "JARs\\lucene-core-3.6.0.jar;");
-	strcat(jarPath, appPath);
-	strcat(jarPath, "xerces.jar;");
-	strcat(jarPath, appPath);
-	strcat(jarPath, "plugin.jar;");
-	strcat(jarPath, appPath);
-	strcat(jarPath, ";");
-	char* envcp = getenv("CLASSPATH");
+	strcpy_s(prefFile, sizeprefFile, bClient ? "sagetvclient " : "sagetv ");
+	strcat_s(prefFile, sizeprefFile, appPath);
+	strcat_s(prefFile, sizeprefFile, bClient ? "SageClient.properties" : "Sage.properties");
+	strcpy_s(jarPath, sizejarPath, "-Djava.class.path=");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "Sage.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "JARs\\lucene-core-3.6.0.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "xerces.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "plugin.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, ";");
+	char* envcp = nullptr;
+	size_t sz = 0;
+	_dupenv_s(&envcp, &sz, "CLASSPATH");
 	if (envcp)
-		strcat(jarPath, envcp);
+		strcat_s(jarPath, sizejarPath, envcp);
 	options[vm_args.nOptions++].optionString = jarPath;
 #endif
 #ifdef SAGE_TV_CLIENT
-	strcpy(prefFile, "sagetvclient ");
-	strcat(prefFile, appPath);
-	strcat(prefFile, "SageClient.properties");
-	strcpy(jarPath, "-Djava.class.path=");
-	strcat(jarPath, appPath);
-	strcat(jarPath, "Sage.jar;");
-	strcat(jarPath, appPath);
-  strcat(jarPath, "JARs\\lucene-core-3.6.0.jar;");
-	strcat(jarPath, appPath);
-	strcat(jarPath, "plugin.jar;");
-	strcat(jarPath, appPath);
-	strcat(jarPath, ";");
-	char* envcp = getenv("CLASSPATH");
+	strcpy_s(prefFile, sizeprefFile, "sagetvclient ");
+	strcat_s(prefFile, sizeprefFile, appPath);
+	strcat_s(prefFile, sizeprefFile, "SageClient.properties");
+	strcpy_s(jarPath, sizejarPath, "-Djava.class.path=");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "Sage.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "JARs\\lucene-core-3.6.0.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, "plugin.jar;");
+	strcat_s(jarPath, sizejarPath, appPath);
+	strcat_s(jarPath, sizejarPath, (";"));
+	char* envcp = nullptr;
+	size_t sz = 0;
+	_dupenv_s(&envcp, &sz, "CLASSPATH");
 	if (envcp)
-		strcat(jarPath, envcp);
+		strcat_s(jarPath, sizejarPath, envcp);

```

Comment:
You need to add a call to free(envcp) as the documentation for _dupenv_s indicates.